scipy/stats/_hypotests.py

```diff
@@ -663,3 +665,217 @@ def somersd(x, y=None):
         raise ValueError("x must be either a 1D or 2D array")
     d, p = _somers_d(table)
     return SomersDResult(d, p, table)
+
+
+def _barnard_exact_compute_gammaln_combination(n):
+    """Compute all log combination of C(n, k)"""
+    gammaln_arr = gammaln(np.arange(n + 1) + 1)
+    return gammaln(n + 1) - gammaln_arr - gammaln_arr[::-1]
+
+
+BarnardExactResult = make_dataclass("BarnardExactResult", [
+    ('statistic', float),
+    ("pvalue", float)
+])
+
+
+def barnard_exact(
+        table,
+        alternative='two-sided',
+        pooled=True,
+        num_it=4):
+    """
+    Perform a Barnard exact test on a 2x2 contingency table.
+
+    Parameters
+    ----------
+    table : array_like of ints
+        A 2x2 contingency table.  Elements should be non-negative integers.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available (default is 'two-sided'):
+
+          * 'two-sided'
+          * 'less': one-sided
+          * 'greater': one-sided
+
+    pooled : bool, optional
+        Whether to compute score statistic with pooled variance or
+        unpooled (Wald statistic)
+
+    num_it : int, optional
+        Number of iterations of the grid search. Default is 4. Must be
+        non-negative.
+
+    Returns
+    -------
+    ber : ``BarnardExactResult`` object
+            The object has attributes ``score_statistic`` and ``p_value``.
+            ``score_statistic`` is the Z statistic with pooled (Score) or
+            unpooled (Wald) variance, depending on the user choice of
+            `pooled` param.
+            ``p_value`` the probability of obtaining a test statistic at least
+            as extreme as the one that was actually observed, assuming that
+            the null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    Notes
+    -----
+    Barnard's test is an exact test used in the analysis of contingency
+    tables. It examines the association of two categorical variables and
+    is a more powerful alternative than Fisher's exact test
+    for 2x2 contingency tables
+
+    References
+    ----------
+    .. [1] `Mehta, Cyrus & Senchaudhuri, Pralay. (2003). Conditional
+    versus Unconditional Exact Tests for Comparing Two Binomials`__.
+
+    .. __: https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf  # noqa: E501
+
```

Comment:
We're getting a doc build error in CI:
```
/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/stats/_hypotests.py:docstring of scipy.stats._hypotests.barnard_exact:61: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/stats/_hypotests.py:docstring of scipy.stats._hypotests.barnard_exact:62: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/circleci/repo/doc/source/generated/scipy.stats.barnard_exact.rst:: WARNING: Anonymous hyperlink mismatch: 0 references but 1 targets.
```
I think this is the problem. I figured you were going for something specific here? But if not we can just make it like other references with URLs:
```suggestion
    .. [1] Mehta, Cyrus and Senchaudhuri, Pralay. (2003). "Conditional
       versus Unconditional Exact Tests for Comparing Two Binomials".
       https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf  # noqa: E501

```